### ISSUE TITLE
Why does shuffling my validation set in Keras change my model's performance?

### ISSUE THREAD

**Issue by MatthewKleinsmith** [fchollet/keras#5162]
**Why I'm confused:**

If I test my model on examples [A, B, C], it will obtain a certain accuracy. If I test the same model on examples [C, B, A], it should obtain the same accuracy. In other words, shuffling the examples shouldn't change my model's accuracy. But that's what seems to be happening below:

**Step-by-step:**

Here is where I train the model:

```python
model.fit_generator(batches, batches.nb_sample, nb_epoch=1, verbose=2,
                    validation_data=val_batches,
                    nb_val_samples=val_batches.nb_sample)
```

Here is where I test the model, without shuffling the validation set:

```python
gen = ImageDataGenerator()
results = []
for _ in range(3):
    val_batches = gen.flow_from_directory(path+"valid", batch_size=batch_size*2,
                                          target_size=target_size, shuffle=False)
    result = model.evaluate_generator(val_batches, val_batches.nb_sample)
    results.append(result)
```

Here are the results (val_loss, val_acc):

```python
[2.8174608421325682, 0.17300000002980231]
[2.8174608421325682, 0.17300000002980231]
[2.8174608421325682, 0.17300000002980231]
```

Notice that the validation accuracies are the same.

Here is where I test the model, with a **shuffled** validation set:

```python
results = []
for _ in range(3):
    val_batches = gen.flow_from_directory(path+"valid", batch_size=batch_size*2,
                                          target_size=target_size, shuffle=True)
    result = model.evaluate_generator(val_batches, val_batches.nb_sample)
    results.append(result)
```

Here are the results (val_loss, val_acc):

```python
[2.8174608802795409, 0.17299999999999999]
[2.8174608554840086, 0.1730000001192093]
[2.8174608268737793, 0.17300000059604645]
```

Notice that **the validation accuracies are inconsistent**, despite an unchanged validation set and an unchanged model. What's going on?

**Comment by bstriner** [comment]
evaluate_generator won't go through every single item in your directory unless the number of samples passed to evaluate_generator is exactly equal to your number of images. Also, there may be some images that are dropped if your count isn't an exact multiple of batch size.

* Can you confirm that val_batches.nb_sample exactly matches the number of images you expect?
* Can you try setting your batch_size to 1 and see if you still see the issue?

Alternatively, if you're using anything like batchnorm, then the differences are expected. If you have different batches, they will have different norms and therefore different results.

If none of the above, it could just be a rounding error or something. The differences are small.

Cheers

**Comment by MatthewKleinsmith** [comment]
> Can you confirm that val_batches.nb_sample exactly matches the number of images you expect?

Yes, the nb_sample attribute from the object returned by flow_from_directory gives this number.

> Can you try setting your batch_size to 1 and see if you still see the issue?

This worked. The accuracies were the same.

However, when I tried a different exact multiple, the inconsistencies came back. I tried a batch size of 100 with 1000 images.

> anything like batchnorm

I'm using VGG16. It doesn't have any batchnorm, but it does have dropout.

> If none of the above, it could just be a rounding error or something. The differences are small.

This might be it. But either way, I'm still confused about why `shuffle=False` seems to prevent the differences while `shuffle=True` seems to cause them.

Thank you

**Comment by MatthewKleinsmith** [comment]
[From StackOverflow](http://stackoverflow.com/a/41841375/3822261):

> This is a really interesting problem. The answer is that it's because of that neural networks are using a float32 format which is not so accurate as float64 - the fluctuation like this are simply the realisation of an underflow phenomenon.
> 
> It case of your loss - you may notice that the differences are occuring after 7th decimal digit of a fractional part - what is exactly the precision of a float32 format. So - basically - you may assume that all numbers presented in your example are equal in terms of a float32 representation.
> \- [Marcin Możejko](http://stackoverflow.com/users/5974433/marcin-mo%c5%bcejko)

My main response:

> Assuming this is an underflow problem, why does setting `shuffle=True` cause the inconsistency while setting `shuffle=False` doesn't?

The rest of my response:

> Anyway, I think you're right. From Wikipedia: "no more than 9 significant decimal digits can be stored" and I see that for the accuracies, the inconsistencies start at the 10th decimal digit. I'm a little confused about the loss values (first column). Why do the inconsistencies start at the 8th decimal digit there but start at the 10th decimal digit for the accuracies (second column)?

To make this super clear:

**Why does setting `shuffle=True` cause the inconsistency while setting `shuffle=False` doesn't?**

**Comment by MatthewKleinsmith** [comment]
I'm closing this issue because my question is answered. I have a new question, but I'll open a new issue for it.

**The question:**

> Why does shuffling my validation set in Keras change my model's performance?


**The answer:**

> It doesn't. Your model's performance is the same.
